File: client-sdks/sockudo-kotlin/lib/src/main/kotlin/io/sockudo/client/MessageDeduplicator.kt
package io.sockudo.client

/**
 * LRU-based message deduplication using a [LinkedHashMap] with access-order eviction.
 * When the cache exceeds [capacity], the least-recently-used entry is evicted.
 */
internal class MessageDeduplicator(private val capacity: Int = 1000) {

    private val seen: LinkedHashMap<String, Boolean> = object : LinkedHashMap<String, Boolean>(
        /* initialCapacity = */ (capacity * 4 / 3) + 1,
        /* loadFactor = */ 0.75f,
        /* accessOrder = */ true,
    ) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Boolean>?): Boolean =
            size > capacity
    }

    /**
     * Returns `true` if [messageId] has already been seen (i.e. is a duplicate).
     */
    fun isDuplicate(messageId: String): Boolean = seen.containsKey(messageId)

    /**
     * Records [messageId] so future calls to [isDuplicate] will return `true`.
     */
    fun track(messageId: String) {
        seen[messageId] = true
    }
}

File: client-sdks/sockudo-swift/Sources/SockudoSwift/MessageDeduplicator.swift
import Foundation

/// Tracks recently seen message IDs to skip duplicate deliveries.
/// Uses an insertion-ordered set with LRU eviction when capacity is exceeded.
final class MessageDeduplicator: @unchecked Sendable {
  private let capacity: Int
  private var order: [String] = []
  private var seen: Set<String> = []

  init(capacity: Int = 1000) {
    precondition(capacity > 0, "Deduplication capacity must be positive")
    self.capacity = capacity
    order.reserveCapacity(capacity)
  }

  /// Returns `true` if this message ID has already been tracked.
  func isDuplicate(_ messageId: String) -> Bool {
    seen.contains(messageId)
  }

  /// Records a message ID. If the cache is at capacity the oldest entry is evicted.
  func track(_ messageId: String) {
    guard seen.contains(messageId) == false else { return }
    if order.count >= capacity {
      let evicted = order.removeFirst()
      seen.remove(evicted)
    }
    order.append(messageId)
    seen.insert(messageId)
  }

  /// Removes all tracked message IDs.
  func reset() {
    order.removeAll(keepingCapacity: true)
    seen.removeAll(keepingCapacity: true)
  }
}
